Add a C# client supporting Tanks over UDP

hchokshi opened the pull request:
To use it, just create a solution (or re-use the existing solution provided) and add an executable project that references the `Script Wars Client` class library project. One nuget package will need to be restored (provides a `BinaryReader` and `BinaryWriter` that handle big-endian data).

Framework level used is .NET Standard 1.6 (depends on the open source .NET Core, rather than the proprietary .NET Framework).

The API is modelled to be very close to the Java one as I basically just translated the Java client, but takes advantages of some C# syntactic sugar like getters and setters. Class names are identical and equivalent to their Java counterparts.

**Update:** hold off on merging until I add inline docs for the rest of the code. Got caught up in solving the bug with the endianness that I forgot docstrings aren't done.

hchokshi committed:
Add C# clientsupporting UDP protocol

hchokshi committed:
Add Tank game API for C# client

Brownshome commented:
Thanks for making this, I was trying to find out who you were after the competition.

As for the code, I want to try and keep the client libraries as generic as possible, so I am a little uneasy about having Visual Studio project files in the client distribution.

I am not a C# dev so I am open to ideas about the most general way to do things. I was thinking either some sort of dll that people can import or a simple structure of source files, but I don't know how either of those options work with the libraries required.

Brownshome committed:
Updated build script and site page

hchokshi commented:
The sln/csproj files and nuget are now standard for all .NET stuff (e.g. I used Jetbrains Rider to make this) so it should be compatible across Visual Studio/Rider/MonoDevelop.

Brownshome commented:
Ok, I was unaware. Let me know when you have the docs in a place where you are happy with them.

hchokshi committed:
Add comprehensive inline documentation to C# API

hchokshi commented:
So I've done the documentation, but don't merge because I've found a bug - the `TankApi` class always reads the current position wrong. I used a hex editor to examine the tick data from the server, and it's correct.

![image](https://user-images.githubusercontent.com/10136407/29484391-dd542f38-8511-11e7-9ec0-a8d694b12685.png)

For some reason my code is reading byte `04` twice, as the X and Y position of the tank. If I understand correctly those bytes are: `response`, `is alive` (red), `ammo` (green), `x y` (blue). Correct?

Still working on figuring out why the Coordinates class reads the Y byte twice. Not a fun time :(

Brownshome commented:
Yes, that is correct in terms of data.

Brownshome committed:
This might be the cause of your data issues

Brownshome commented on client/csharp/Script Wars Client/Script Wars Client/ScriptWars/Games/Tanks/World.cs:
```diff
@@ -0,0 +1,108 @@
+            // Wall positions
+            for (var i = 0; i < _map.GetLength(0); i++)
+            for (var j = 0; j < _map.GetLength(1); j++)
+                _map[i, j] = network.ReadBool();
```
As the map is sent row by row, j is the x coordinate. This is probably a bug, as this is inconsistent with the IsWall function.

hchokshi replied:
Yeah, I'll test that again. I thought I was reading this incorrectly and that was causing the coordinate issues.

hchokshi commented on client/csharp/Script Wars Client/Script Wars Client/ScriptWars/Games/Tanks/Coordinates.cs:
```diff
@@ -17,7 +17,7 @@ public class Coordinates
         public Coordinates(int x, int y)
         {
-            X = y;
+            X = x;
```
Wow I'm an idiot. Didn't even consider that I screwed up that constructor instead of something reading into it (which was more likely to be error prone since the read order matters).

Brownshome committed:
Fix small issues with map & boolean reading

Brownshome commented:
I fixed the map issues and a problem with reading booleans in f7fc6958ee1f9e04d47fd9c31fdad82d3707c79c

I have yet to test shot mechanics but map reading seems to be ok now.

hchokshi committed:
Add Random AI starter project to C# client

hchokshi commented:
I've tested this and it's all working now! You can go ahead and merge if it looks good to you.

I also added a C# implementation of the random AI (random movement, shoots at people where possible). This should hopefully serve as a starting point, and offer an example of making a separate AI project in the solution that references `Script Wars Client` as a dependency.

hchokshi committed:
Use named Width/Height for loops populating map

Makes it more clear which loop variable acts in which axis.

Brownshome committed:
Added the GetShot method

hchokshi committed:
Store map as transpose to optimise locality of reference

Storing width on the first axis means switching outer dimension every
time we write a bool to the map array. This is inefficient as it has
poor locality of reference. Storing as tranpose will reduce cache
misses while populating the array.

Brownshome commented:
Alright, everything looks good to go, thank you very much. I'll make sure to add your name to the credits list. Do you have anything you want to add before I merge?

hchokshi commented:
Nope, it's pretty good and done imo! :D

Brownshome commented:
Maybe I'll try and persuade the gradle build script into making a pre-compiled version as well. But for now I think the source is enough. Thank you muchly.
